tests/regression/src/lib.rs

```diff
@@ -87,70 +93,202 @@ mod tests {
+        print!("{prev_file} and {curr_file}");
+        let diff_output = run_cg_annotate_diff(&prev_file, &curr_file);
+        save_diff_output(&diff_output, test_name);
+
+        let diff = find_instruction_count(&diff_output)
+            .expect("Failed to parse cg_annotate --diff output");
+
+        assert_diff_within_threshold(diff, test_name);
+    }
+
+    fn find_and_validate_diff_files(test_name: &str) -> (String, String) {
+        let annotated_files = find_annotated_files(test_name);
+        let file_len = annotated_files.len();
+        if file_len != 2 {
+            panic!("Expected exactly 2 annotated files for {test_name}, found {file_len}");
+        }
+
+        let file1 = &annotated_files[0];
+        let file2 = &annotated_files[1];
+
+        if !is_commit_in_log(file1) || !is_commit_in_log(file2) {
+            let first_hash = extract_commit_hash(file1);
+            let second_hash = extract_commit_hash(file2);
+            panic!(
+                "One or both commit hashes are not in the git log: {first_hash} or {second_hash}"
+            );
+        }
+
+        let (old_file, new_file) = if is_older_commit(file1, file2) {
+            (file1.clone(), file2.clone())
+        } else if is_older_commit(file2, file1) {
+            (file2.clone(), file1.clone())
+        } else {
+            panic!("Cannot determine the older commit between {file1} and {file2}",);
+        };
+
+        (old_file, new_file)
+    }
+
+    fn find_annotated_files(test_name: &str) -> Vec<String> {
+        let pattern = format!("target/cg_artifacts/cachegrind_{test_name}_*.out");
+        glob::glob(&pattern)
+            .expect("Failed to read glob pattern")
+            .filter_map(Result::ok)
+            .map(|path| path.to_string_lossy().into_owned())
+            .collect()
+    }
+
+    fn is_commit_in_log(file: &str) -> bool {
+        let commit = extract_commit_hash(file);
+        let output = Command::new("git")
+            .args(["log", "--pretty=format:%H"])
+            .output()
+            .expect("Failed to execute git log");
+        let log = String::from_utf8(output.stdout).expect("Invalid UTF-8 in git log output");
+        log.lines().any(|line| line == commit)
+    }
+
+    fn is_older_commit(file1: &str, file2: &str) -> bool {
+        let commit1 = extract_commit_hash(file1);
+        let commit2 = extract_commit_hash(file2);
+        let output = Command::new("git")
+            .args(["merge-base", "--is-ancestor", &commit1, &commit2])
+            .status()
+            .expect("Failed to execute git merge-base");
+        output.success()
     }
 
-    /// Parses the annotated file for the overall instruction count total
-    fn find_instruction_count(file_path: &str) -> Result<u64, io::Error> {
-        let path = Path::new(file_path);
-        let file = File::open(path)?;
-        let reader = io::BufReader::new(file);
+    fn extract_commit_hash(file: &str) -> String {
```

Comment:
For these kinds of methods I always like to include an example of the `input` and the output. So just add a comment with
```
// target/cg_artifacts/cachegrind_foo_test_$COMMITHASH.out
```

I also this is method would be a bit simpler if the commit hash was the folder name instead of a substring of the file name.